### ISSUE TITLE
Analog to TfIdfVectorizer

### ISSUE THREAD

**Issue by ClaytonJY** [EmilHvitfeldt/textrecipes#3]
@EmilHvitfeldt thanks for starting this; got here from https://github.com/tidymodels/recipes/issues/192.

I've been meaning to write a `step_tfidf` that does something roughly equivalent to [`sklearn`s `TfIdfVectorizer`](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html), though hopefully without quite so many arguments. This package's existence seems like a good excuse to finally do that.

So my questions then are:
1. are you open to a PR for this, or do you already have something in the works?
1. all-in-one, or layered? All-in-one would mean the input to this step is one row per document (e.g. a `doc` columns with long-form text in it), and layered meaning this would need to be called _after_ `step_tokenize`. I think I lean towards the layered approach, because otherwise there'd be a ton of duplication between the two, but all-in-one might be more user-friendly? One argument _against_ the layers is that the output of `step_tokenizer` will only ever be an intermediate one (can't model from the output, right?), so perhaps it should be a functional input to other steps, rather than it's own step?
1. Should filtering (like with the `max_df`/`min_df` args in sklearn) be done in this step, or in a subsequent step? Something seems lovely and idiomatic about the layered approach, but filtering on document frequency after-the-fact would be way less efficient.

**Comment by EmilHvitfeldt** [comment]
Hello @ClaytonJY! Glad you could find it.

1. I'm waiting for some design decisions from #1, but I do have something in the works :)

2. I'm going for the layered approach. This way you get more flexibility and you avoid having a step with a million arguments. Unfortunately you will get a handful of intermediate steps, but that is a downside I'm willing to take and will counteract with appropriate documentation. 

3. filtering should be a step by itself as it should be able to be place before or after `step_stem` (or similar steps) depending on the approach. (I'm not saying that one order is better then the order, but as far as I know none of the orders are superior 100% of the times.)

My goal would be having something like

```r
recipe(label ~ text, data = data) %>%
  step_tokenize("word") %>%
  step_stem() %>%
  step_textfilter(max_features = 5000) %>%
  step_tfidf()

# Or

recipe(label ~ text, data = data) %>%
  step_tokenize("word") %>%
  step_text_filter(max_features = 5000) %>%
  step_stem() %>%
  step_tfidf()
```

**Comment by ClaytonJY** [comment]
Okay cool; I follow you on all counts. I like where you're going with 3, the flexibility seems nice; stem-then-filter, or filter-then-stem.

Something I worry a bit about is that computing term-frequencies will be done at least twice, once in `step_textfilter` and again in `step_tfidf`, which could be a bit painful for a large enough corpus. You might be able to save them inside the recipe, and also have step_stem adjust them accordingly? Seems a bit tricky, but not impossible.

Do any existing `recipes` steps rely on introspecting _inside_ the input recipe (e.g. reading params trained in earlier step), rather than just the output of the step? I think we'd want that ability, so filtering on doc-frequencies chould be post-`step_tfidf` and use the values computed and saved inside `tf_idf` (necessary to keep DF's for applying to test set) which won't actually be exposed in the output.

**Comment by topepo** [comment]
> Do any existing recipes steps rely on introspecting inside the input recipe (e.g. reading params trained in earlier step), rather than just the output of the step?

No. That's going to happen eventually but would be a big api change. We would need to label the steps (or have some other method of reference them). We also need to have a way of undoing some steps (like uncenter and unscale) so there is the need for this. It's not close on my to-do list though. If you have a strategy, then let me know. 

There might be the need to check for tokenization in other steps (where that format is required) and I think that an initial step there is to throw an error of that's not the case. We can change that later, but it might be initial idea. 

The input/output table in the other issue would be a big help here. 

`step_tfidf` is going to make to save the training set word frequencies in order to calculate on new samples (or something to that effect).

**Comment by ClaytonJY** [comment]
> No. That's going to happen eventually but would be a big api change. We would need to label the steps (or have some other method of reference them). We also need to have a way of undoing some steps (like uncenter and unscale) so there is the need for this. It's not close on my to-do list though. If you have a strategy, then let me know.

I've got some starting thoughts (mostly "S3 is probably not formal enough"), is there a good `recipes` issues to discuss this on?

> step_tfidf is going to make to save the training set word frequencies in order to calculate on new samples (or something to that effect).

Document frequencies, not term frequencies, unless TF's are needed for filtering. Something to consider @EmilHvitfeldt is how to save IDF's; the usual logged-ratio wouldn't easily support nice filtration, but anything else pushes computation elsewhere & repeats it. I suppose first-pass could just be fully-computed IDF's, and that can be changed later without changing the external API.

**Comment by EmilHvitfeldt** [comment]
tfidf calculation are now supported via the [text2vec](https://github.com/dselivanov/text2vec) package.
